In [8]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Dropout
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.regularizers import l2
from tensorflow.keras.layers import Conv2D, MaxPooling2D

from tensorflow.keras.metrics import Precision, Recall, AUC

from tensorflow.keras.utils import to_categorical


# Завантаження датасету fashion_mnist

In [17]:
fashion_mnist = tf.keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Нормалізація даних

In [18]:

train_images = train_images / 255.0
test_images = test_images / 255.0
train_labels_one_hot = to_categorical(train_labels)
test_labels_one_hot = to_categorical(test_labels)

#Модель

In [23]:
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D((2, 2)),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Conv2D(64, (3, 3), activation='relu'),
    Flatten(),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy', Precision(), Recall(), AUC()])

history = model.fit(train_images, train_labels_one_hot, epochs=13, batch_size=64, validation_split=0.2)



Epoch 1/13
750/750 [==============================] - 45s 58ms/step - loss: 0.5980 - accuracy: 0.7821 - precision_10: 0.8635 - recall_10: 0.7063 - auc_8: 0.9792 - val_loss: 0.4087 - val_accuracy: 0.8583 - val_precision_10: 0.8948 - val_recall_10: 0.8192 - val_auc_8: 0.9899
Epoch 2/13
750/750 [==============================] - 43s 58ms/step - loss: 0.3644 - accuracy: 0.8688 - precision_10: 0.8952 - recall_10: 0.8439 - auc_8: 0.9914 - val_loss: 0.3445 - val_accuracy: 0.8781 - val_precision_10: 0.9054 - val_recall_10: 0.8528 - val_auc_8: 0.9925
Epoch 3/13
750/750 [==============================] - 43s 58ms/step - loss: 0.3098 - accuracy: 0.8877 - precision_10: 0.9074 - recall_10: 0.8682 - auc_8: 0.9934 - val_loss: 0.3073 - val_accuracy: 0.8881 - val_precision_10: 0.9086 - val_recall_10: 0.8694 - val_auc_8: 0.9938
Epoch 4/13
750/750 [==============================] - 43s 57ms/step - loss: 0.2781 - accuracy: 0.8979 - precision_10: 0.9153 - recall_10: 0.8840 - auc_8: 0.9945 - val_loss: 0.298

# Оцінка моделі на тестових даних

In [27]:
test_labels_one_hot = to_categorical(test_labels)
test_loss, test_accuracy, test_precision, test_recall, test_auc = model.evaluate(test_images[-1500:], test_labels_one_hot[-1500:])
# Print the metrics
print(f"Accuracy: {test_accuracy * 100:.2f}%")
print(f"Precision: {test_precision:.2f}")
print(f"Recall: {test_recall:.2f}")
print(f"AUC: {test_auc:.2f}")

47/47 [==============================] - 0s 10ms/step - loss: 0.2723 - accuracy: 0.9127 - precision_10: 0.9220 - recall_10: 0.9067 - auc_8: 0.9933
Accuracy: 91.27%
Precision: 0.92
Recall: 0.91
AUC: 0.99


Узагальнюючи отримані результати, можна сказати, що розроблена нейронна мережа виявилася дуже ефективною у вирішенні задачі класифікації об'єктів набору даних Fashion MNIST. Значення точності 91.27% свідчить про високу здатність моделі правильно класифікувати різні категорії одягу. Також варто відзначити високі показники точності (Precision) та повноти (Recall), які становлять 0.92 та 0.91 відповідно. Це означає, що модель не тільки часто робить правильні прогнози, але й рідко пропускає правильні класи, а також має низьку ймовірність помилкового позитивного результату.

Особливо вражає висока площа під кривою характеристик роботи приймача (AUC), яка дорівнює 0.99. Це свідчить про те, що модель має відмінну здатність розрізняти між різними класами, що є критично важливим у багатьох застосуваннях, де необхідно мінімізувати помилки.